autoPyTorch/pipeline/components/training/trainer/base_trainer_choice.py

```diff
@@ -0,0 +1,479 @@
+            if self.early_stop_handler(X):
+                break
+
+            if self.choice.on_epoch_end(X=X, epoch=epoch):
+                break
+
+            self.logger.debug(self.run_summary.repr_last_epoch())
+
+            # Reached max epoch on next iter, don't even go there
+            if budget_tracker.is_max_epoch_reached(epoch + 1):
+                break
+
+            epoch += 1
+
+            torch.cuda.empty_cache()
+
+        # wrap up -- add score if not evaluating every epoch
+        if not self.eval_valid_each_epoch(X):
+            val_loss, val_metrics = self.choice.evaluate(X['val_data_loader'])
+            if 'test_data_loader' in X and X['val_data_loader']:
+                test_loss, test_metrics = self.choice.evaluate(X['test_data_loader'])
+            self.run_summary.add_performance(
+                epoch=epoch,
+                start_time=start_time,
+                end_time=time.time(),
+                train_loss=train_loss,
+                val_loss=val_loss,
+                test_loss=test_loss,
+            )
+            self.logger.debug(self.run_summary.repr_last_epoch())
+            self.save_model_for_ensemble()
+
+        self.logger.info(f"Finished training with {self.run_summary.repr_last_epoch()}")
+
+        # Tag as fitted
+        self.fitted_ = True
+
+        return X['network'].state_dict()
+
+    def early_stop_handler(self, X: Dict[str, Any]) -> bool:
+        """
+        If early stopping is enabled, this procedure stops the training after a
+        given patience
+        Args:
+            X (Dict[str, Any]): Dictionary with fitted parameters. It is a message passing
+                mechanism, in which during a transform, a components adds relevant information
+                so that further stages can be properly fitted
+
+        Returns:
+            bool: If true, training should be stopped
+        """
+        assert self.run_summary is not None
+        bad_epochs = self.run_summary.get_best_epoch() - self.run_summary.get_last_epoch()
+        if bad_epochs > X['early_stopping']:
+            return True
+
+        return False
+
+    def eval_valid_each_epoch(self, X: Dict[str, Any]) -> bool:
+        """
+        Returns true if we are supposed to evaluate the model on every epoch,
+        on the validation data. Usually, we only validate the data at the end,
+        but in the case of early stopping, is appealing to evaluate each epoch.
+        Args:
+            X (Dict[str, Any]): Dictionary with fitted parameters. It is a message passing
+                mechanism, in which during a transform, a components adds relevant information
+                so that further stages can be properly fitted
+
+        Returns:
+            bool: if True, the model is evaluated in every epoch
+
+        """
+        if 'early_stopping' in X and X['early_stopping']:
+            return True
+
+        # We need to know if we should reduce the rate based on val loss
```

Comment:
Not strictly, that could also be done when evaluating every e.g. 5 epochs. It would be nice if we could support that

Reply:
I can just code something quick to support this, but I will rather add proper checking and support for this, and hence I created https://github.com/LMZimmer/Auto-PyTorch_refactor/issues/24